src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportJobTask.cs

```diff
@@ -354,6 +416,69 @@ private async Task<IAnonymizer> CreateAnonymizerAsync(CancellationToken cancella
                 queryParametersList.Add(Tuple.Create(KnownQueryParameterNames.Type, _exportJobRecord.ResourceType));
             }
 
+            var filteredResources = new HashSet<string>();
+            if (_exportJobRecord.Filters != null)
+            {
+                foreach (var filter in _exportJobRecord.Filters)
+                {
+                    filteredResources.Add(filter.Type);
+                }
+            }
+
+            if (!progress.RunFilteredSearches)
+            {
+                var index = -1;
+                if (_exportJobRecord.Filters != null)
+                {
+                    foreach (var filter in _exportJobRecord.Filters)
+                    {
+                        index++;
+                        if (progress.CurrentFilter != null && filter != progress.CurrentFilter)
+                        {
+                            continue;
+                        }
+
+                        if (_exportJobRecord.ExportType == ExportJobType.All || filter.Type.Equals(KnownResourceTypes.Patient, StringComparison.Ordinal))
+                        {
+                            progress.SetFilter(filter);
+
+                            List<Tuple<string, string>> filterQueryParametersList = new List<Tuple<string, string>>(queryParametersList);
+                            filterQueryParametersList.Add(Tuple.Create(KnownQueryParameterNames.Type, filter.Type));
+                            foreach (var param in filter.Parameters)
+                            {
+                                filterQueryParametersList.Add(param);
+                            }
+
+                            await SearchCompartmentWithFilter(exportJobConfiguration, progress, filterQueryParametersList, batchIdPrefix + "-filter" + index, cancellationToken);
+                        }
+                    }
+                }
+
+                progress.MarkFiltersFinished();
+                await UpdateJobRecordAsync(cancellationToken);
+            }
+
+            if (_exportJobRecord.ExportType == ExportJobType.All || !filteredResources.Contains(KnownResourceTypes.Patient))
```

Comment:
If there are 2 filters - one for `Patient` and another for `MedicationRequest` - it looks like we will never use the `MedicationRequest` filter while searching the compartment. `!filteredResources.Contains(KnownResourceTypes.Patient)` will return true and the previous if condition runs the search only with `Patient` filters.

Reply:
Wow, this was really broken. Thank you for catching this. I'll fix this and add some new unit tests.